contents/java/JUnit.md

```diff
@@ -0,0 +1,243 @@
+<frontmatter>
+  title: JUnit
+  footer: footer.md
+  head: head.md
+  siteNav: mainNav.md
+  pageNav: 3
+</frontmatter>
+
+{{ navbar | safe }}
+
+<div class="website-content">
+
+# JUnit
+
+<b>Author: [Lin Si Jie](https://github.com/sijie123)</b>
+
+Reviewers: [Aadyaa Maddi](https://github.com/amad-person), [Marvin Chin](https://github.com/marvinchin)
+
+## What is JUnit?
+_____
+
+**JUnit is an _automated testing framework_ for Java** i.e., it can be used to programmatically verify the actual behavior of Java code against the expected behavior.
+
+Let's use a simple example to see how JUnit is used. Suppose we want to test the method `calculateArraySum(int[] values, int n)` which returns the sum of the first `n` elements of `values`.
+
+```java
+public class Utils {
+  // Sum up first n elements in values, and return the result
+  public static int calculateArraySum(int[] values, int n) {
+    int sum = 0;
+    for (int i = 1; i < n; i++) {
+      sum += values[i];
+    }
+    return sum;
+  }
+}
+```
+
+Let's write a test to check the correctness of our code. This test will use a sample array of size 5 that sums to 15, and check that our `calculateArraySum` returns the correct result.
+
+```java
+public class UtilsTest {
+  @Test
+  public void calculateArraySum_fiveNumbers_correctAnswer() {
+    int[] testArray = {1, 2, 3, 4, 5};
+    assertEquals(15, calculateArraySum(testArray, 5));
+  }
+}
+```
+<tip-box>:bulb: By convention, we append `Test` to the name of the tested class. It is also good practice to write a descriptive name for the test, so other developers will know the intention and scope of each test.</tip-box>
+
+That's it! We are ready to run the tests through our IDE. The `@Test` tag tells JUnit that the `calculateArraySum_fiveNumbers_correctAnswer` method is meant to be a test. JUnit will then automatically execute our test(s).
+
+Let's have a look at the test results. 
+
+<img src="{{baseUrl}}/contents/java/assertFailed.PNG" />
+
+Hmm, the test seems to fail when we run it.
+Did you notice the subtle bug in `calculateArraySum`?
+
+
+We "forgot" to include the 0-th element of inputArray, a common off-by-one bug.
+With JUnit, we can use automatic tests to ensure that our functions are correct. This is easier and less error-prone than repeatedly testing and calculating manually.
+
+Now, let's fix the code.
+
+```java
+public class Utils {
+  // Sum up first n elements in values, and return the result
+  public static int calculateArraySum(int[] values, int n) {
+    int sum = 0;
+    for (int i = 0; i < n; i++) {
+      sum += values[i];
+    }
+    return sum;
+  }
+}
+```
+
+<img src="{{baseUrl}}/contents/java/assertPass.PNG" />
+
+Running the same unit test again, we now have a passing test. Our `calculateArraySum` function works as intended.
+
+<p></p> 
+
+## Why JUnit?
+
+_____
```

Comment:
Let's not add these horizontal lines for level 2 headings. In other articles, only level 1 headings have such horizontal lines (auto-added).